### DISCUSSION TITLE
Issues with CsvDataLoader – Cannot Bypass Despite Column Verification

### DISCUSSION THREAD

**Discussion by NituSidhu** [google/meridian#337] [Q&A]
I’m encountering a persistent issue with the CsvDataLoader step when attempting to load my dataset, and I can’t seem to bypass it despite verifying my columns. I’ve followed the documentation and double-checked my column mappings, but I continue to receive errors. Here’s a breakdown of the setup and the troubleshooting steps I’ve attempted so far:

CsvDataLoader Setup:

```
python
Copy code
loader = load.CsvDataLoader(
    csv_path="raw_data.csv",
    kpi_type='non_revenue',
    coord_to_columns=coord_to_columns,
    media_to_channel=correct_media_to_channel,
    media_spend_to_channel=correct_media_spend_to_channel,
)
data = loader.load()
```
coord_to_columns Configuration: I have set up coord_to_columns according to the documentation, ensuring all necessary fields are either populated with column names from my dataset or set to None where appropriate:
```
coord_to_columns = load.CoordToColumns(
    time='Date',
    geo=None,
    controls='googletrends_keyword',
    population=None,
    kpi='conversions',
    revenue_per_kpi=None,  # Optional
    media=[
        'dsp_i', 'meta_i', 'reddit_i', 'sem_i', 
        'podcast_i', 'tiktok_i', 'twitter_i'
    ],
    media_spend=[
        'dsp_s', 'meta_s', 'reddit_s', 'sem_s', 
        'podcast_s', 'tiktok_s', 'twitter_s'
    ],
)
```
Media Channel Mapping: My media and spend mappings align with the channel naming convention suggested:

```
correct_media_to_channel = {
    'dsp_i': 'off_dsp',
    'meta_i': 'off_meta',
    # ... (other mappings)
}
correct_media_spend_to_channel = {
    'dsp_s': 'dsp',
    'meta_s': 'meta',
    # ... (other mappings)
}
```

Column Verification: I’ve verified that all columns in coord_to_columns, correct_media_to_channel, and correct_media_spend_to_channel exist in raw_data.csv. Additionally, I am not attempting to use every column in the CSV.

Error Details: Despite these efforts, I continue to encounter an error related to NoneType and str when loading the data, and I am unable to bypass or troubleshoot further without additional guidance.

```
/usr/local/lib/python3.10/dist-packages/meridian/data/load.py:770: UserWarning: The `population` argument is ignored in a nationally aggregated model. It will be reset to [1, 1, ..., 1]
  warnings.warn(
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
[<ipython-input-56-aaafa99514a7>](https://localhost:8080/#) in <cell line: 1>()
----> 1 loader = load.CsvDataLoader(
      2     csv_path='/content/raw_data.csv',
      3     kpi_type='non_revenue',
      4     coord_to_columns=coord_to_columns,
      5     media_to_channel=correct_media_to_channel,

3 frames
[/usr/local/lib/python3.10/dist-packages/meridian/data/load.py](https://localhost:8080/#) in _validate_column_names(self)
    736     desired_columns = sorted(desired_columns)
    737 
--> 738     actual_columns = sorted(self.df.columns.to_list())
    739     if any(d not in actual_columns for d in desired_columns):
    740       raise ValueError(

TypeError: '<' not supported between instances of 'NoneType' and 'str'

```

I'm curious if anyone can provide insights or advice on how to not run into any TypeError messages, or if there are any specific requirements for optional columns that may be affecting this?

**Comment by NituSidhu** [comment]
Moving this question as an Issue instead: https://github.com/google/meridian/issues/343

**Comment by Jon77Ruler** [comment]
```
correct_media_to_channel = {
    'dsp_i': 'off_dsp',
    'meta_i': 'off_meta',
    # ... (other mappings)
}
correct_media_spend_to_channel = {
    'dsp_s': 'dsp',
    'meta_s': 'meta',
    # ... (other mappings)
}
```

In the correct media spend to channel, shouldn't the entries on the right match those in section above.

i.e. `'dsp_s': 'off_dsp'`  and `'metai_i': 'off_meta'`

etc. ?
